metrics/query.global/global.go

```diff
@@ -0,0 +1,194 @@
+			OPT_RESPONSE_TIME_OPTIONAL: {
+				Name:    OPT_RESPONSE_TIME_OPTIONAL,
+				Desc:    "If collecting response time metrics is optional. This will fail if it is required but not available",
+				Default: "yes",
+				Values: map[string]string{
+					"yes": "Optional",
+					"no":  "Required",
+				},
+			},
+			OPT_RESPONSE_TIME_PERCENTILES: {
+				Name:    OPT_RESPONSE_TIME_PERCENTILES,
+				Desc:    "Comma-separated list of response time percentiles formatted as 999, 0.999 or 99.9",
+				Default: DEFAULT_PERCENTILE_OPTION,
+				//Values:  map[string]string{},
+			},
+			OPT_TRUNCATE_TABLE: {
+				Name:    OPT_TRUNCATE_TABLE,
+				Desc:    "If the source table should be truncated to reset data after each retrieval",
+				Default: "yes",
+				Values: map[string]string{
+					"yes": "Truncate source table after each retrieval",
+					"no":  "Do not truncate source table after each retrieval",
+				},
+			},
+		},
+		Metrics: []blip.CollectorMetric{
+			{
+				Name: "response_time",
+				Type: blip.GAUGE,
+				Desc: "Query response time (microseconds)",
+			},
+		},
+	}
+}
+
+// Prepare prepares the collector for the given plan.
+func (c *Global) Prepare(ctx context.Context, plan blip.Plan) (func(), error) {
+	_, err := c.db.Query(BASE_QUERY)
+	if err != nil {
+		blip.Debug("response time metric is not available: %v", err.Error())
+		c.available = false
+	}
+
+LEVEL:
+	for _, level := range plan.Levels {
+		dom, ok := level.Collect[DOMAIN]
+		if !ok {
+			continue LEVEL // not collected at this level
+		}
+
+		err := c.prepareLevel(dom, level)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// Collect collects metrics at the given level.
+func (c *Global) Collect(ctx context.Context, levelName string) ([]blip.MetricValue, error) {
+	if !c.available {
+		if !c.optional[levelName] {
+			panic(fmt.Sprintf("%s: required response time metrics couldn't be collected", levelName))
+		} else {
+			return nil, nil
+		}
+	}
+
+	var metrics []blip.MetricValue
+	for _, percentile := range c.percentiles[levelName] {
+		where := fmt.Sprintf(" WHERE bucket_quantile >= %f ORDER BY bucket_quantile LIMIT 1", percentile)
+		query := BASE_QUERY + where
+
+		var p float64
+		var us float64
+		err := c.db.QueryRowContext(ctx, query).Scan(&p, &us)
+		if err != nil {
+			return nil, fmt.Errorf("%s: error running query for response time metric: %s", levelName, err)
+		}
+
+		m := blip.MetricValue{
+			Type:  blip.GAUGE,
+			Name:  "response_time",
+			Value: us,
+			Meta: map[string]string{
+				util.FormatPercentile(percentile): fmt.Sprintf("%.1f", p),
+			},
+		}
+		metrics = append(metrics, m)
+	}
+
+	if c.truncateTable[levelName] {
+		_, err := c.db.Exec(TRUNCATE_QUERY)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return metrics, nil
+}
+
+// prepareLevel sanitizes options for given level based on user specified options
+func (c *Global) prepareLevel(dom blip.Domain, level blip.Level) error {
```

Comment:
Let's move this into `Prepare` because it is the heart of that func. Factoring it out (in this case) adds no value; it just splits up a func's main code, which leads the reader around needlessly.